presto-main/src/main/java/com/facebook/presto/util/FastutilSetHelper.java

```diff
@@ -13,51 +13,52 @@
-import it.unimi.dsi.fastutil.chars.CharOpenHashSet;
-import it.unimi.dsi.fastutil.doubles.DoubleOpenHashSet;
-import it.unimi.dsi.fastutil.floats.FloatOpenHashSet;
-import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
-import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+import it.unimi.dsi.fastutil.doubles.DoubleHash;
+import it.unimi.dsi.fastutil.doubles.DoubleOpenCustomHashSet;
+import it.unimi.dsi.fastutil.longs.LongHash;
+import it.unimi.dsi.fastutil.longs.LongOpenCustomHashSet;
+import it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet;
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
-import it.unimi.dsi.fastutil.shorts.ShortOpenHashSet;
 
+import java.lang.invoke.MethodHandle;
 import java.util.Collection;
 import java.util.Set;
 
+import static com.facebook.presto.metadata.OperatorType.EQUAL;
+import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
+import static com.facebook.presto.spi.StandardErrorCode.INTERNAL_ERROR;
+
 public final class FastutilSetHelper
 {
     private FastutilSetHelper() {}
 
     @SuppressWarnings({"unchecked"})
-    public static Set<?> toFastutilHashSet(Set<?> set, Class<?> javaElementType)
+    public static Set<?> toFastutilHashSet(Set<?> set, Type type, FunctionRegistry registry)
     {
-        if (javaElementType == boolean.class) {
-            return new BooleanOpenHashSet((Collection<Boolean>) set);
-        }
-        if (javaElementType == byte.class) {
-            return new ByteOpenHashSet((Collection<Byte>) set);
-        }
-        if (javaElementType == char.class) {
-            return new CharOpenHashSet((Collection<Character>) set);
+        Class<?> javaElementType = type.getJavaType();
+        if (javaElementType == long.class) {
+            return new LongOpenCustomHashSet((Collection<Long>) set, new LongStrategy(registry, type));
         }
         if (javaElementType == double.class) {
-            return new DoubleOpenHashSet((Collection<Double>) set);
-        }
-        if (javaElementType == float.class) {
-            return new FloatOpenHashSet((Collection<Float>) set);
-        }
-        if (javaElementType == int.class) {
-            return new IntOpenHashSet((Collection<Integer>) set);
+            return new DoubleOpenCustomHashSet((Collection<Double>) set, new DoubleStrategy(registry, type));
         }
-        if (javaElementType == long.class) {
-            return new LongOpenHashSet((Collection<Long>) set);
+        if (javaElementType == boolean.class) {
+            return new BooleanOpenHashSet((Collection<Boolean>) set);
         }
-        if (javaElementType == short.class) {
-            return new ShortOpenHashSet((Collection<Short>) set);
+        if (javaElementType == Slice.class) {
+            return new ObjectOpenHashSet<>((Collection<Slice>) set);
         }
-        return new ObjectOpenHashSet(set);
+
+        return new ObjectOpenCustomHashSet(set, new ObjectStrategy(registry, type));
```

Comment:
Return object set only if the type is not a primitive as you did in another pull request.